# PixelDrop — Pandas Project
## Notebook 01: Load & Explore Raw Data

**Goal:** Load all 5 CSV files and understand what we're working with before cleaning anything.

**Tables:**
- `customers` — 4,500 customer profiles
- `products` — 15 products in the catalogue
- `orders` — 12,000 order headers
- `order_items` — 29,037 line items
- `returns` — 1,500 return records

## Step 1: Import Libraries
We import all the tools we need at the top of every notebook.
- `pandas` → load, clean, analyse data
- `numpy` → numbers and math
- `matplotlib` → charts
- `seaborn` → prettier charts

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print('Libraries loaded successfully!')

Libraries loaded successfully!


## Step 2: Load the Data
Load all 5 CSVs into pandas DataFrames.
Each CSV becomes a DataFrame — think of it like a table in SQL.

> **Update the path below to match your desktop folder path!**

In [2]:
# Update this path to your claude folder
PATH = r'C:\Users\\איתי\Desktop\claude\data\raw\\'

customers   = pd.read_csv(PATH + 'customers.csv')
products    = pd.read_csv(PATH + 'products.csv')
orders      = pd.read_csv(PATH + 'orders.csv')
order_items = pd.read_csv(PATH + 'order_items.csv')
returns     = pd.read_csv(PATH + 'returns.csv')

print('All 5 files loaded successfully!')

All 5 files loaded successfully!


## Step 3: Check Shape of All Tables
`.shape` returns `(rows, columns)` — like `SELECT COUNT(*)` in SQL.

In [3]:
tables = {
    'customers': customers,
    'products': products,
    'orders': orders,
    'order_items': order_items,
    'returns': returns
}

print('Table shapes:')
print('-' * 35)
for name, df in tables.items():
    print(f'{name:<15} {df.shape[0]:>6} rows  {df.shape[1]} columns')

Table shapes:
-----------------------------------
customers         4500 rows  7 columns
products            15 rows  7 columns
orders           12000 rows  8 columns
order_items      29037 rows  6 columns
returns           1500 rows  7 columns


## Step 4: Preview Each Table
`.head()` shows the first 5 rows — like `SELECT TOP 5` in SQL.

In [4]:
print('=== CUSTOMERS ===')
display(customers.head())

=== CUSTOMERS ===


,customer_id,customer_name,email,phone,state,signup_date,segment
0,CUST-1000,grace young,youngg@hotmail.com,(993) 563-2729,mi,2021/07/03,Occ
1,CUST-1001,Ava Scott,ava_scott@outlook.com,(052) 327-9282,AZ,2023/08/05,V.I.P
2,CUST-1002,Aria K.,kinga@gmail.com,+14862570292,Ariz.,2022-02-16,Occ
3,CUST-1003,sophia moore,moores@hotmail.com,+18409053665,virginia,29-08-2021,New
4,CUST-1004,Lily Gonzalez,lily_gonzalez@outlook.com,6471764184,florida,05-08-2021,new


In [5]:
print('=== PRODUCTS ===')
display(products.head())

=== PRODUCTS ===


,product_id,product_name,category,cost_price,retail_price,stock_qty,description
0,PROD-01,AIR HOODIE,HOODIES,45.0,89.99,320.0,NaN
1,PROD-02,Classic T,Tees,10.0,29.99,850.0,Premium Classic Tee from PixelDrop.
2,PROD-03,Drrop Tee,Tees,12.0,34.99,600.0,NaN
3,PROD-04,CARGO PANTS,bottoms,52.0,119.99,280.0,Premium Cargo Pants from PixelDrop.
4,PROD-05,JOGGER SLIM,BOTTOMS,33.0,79.99,NaN,Premium Jogger Slim from PixelDrop.


In [6]:
print('=== ORDERS ===')
display(orders.head())

=== ORDERS ===


,order_id,order_date,customer_id,shipping_state,payment_method,order_status,shipping_days,notes
0,ORD-10000,10/16/2024,CUST-3992,N.J,Shop Pay,Completed,5.0,NaN
1,ORD-10001,2023-07-12,CUST-1243,georgia,shop pay,REFUNDED,10.0,urgent
2,10002,15-06-2023,CUST-2918,washington,PAYPAL,REFUNDED,13.0,urgent
3,ORD-10003,Sep 25 2024,CUST-4142,north carolina,CREDIT,Completed,2.0,NaN
4,ORD-10004,06-05-2024,CUST-1972,Wash.,PAYPAL,completed,1.0,NaN


In [7]:
print('=== ORDER ITEMS ===')
display(order_items.head())

=== ORDER ITEMS ===


,item_id,order_id,product_id,quantity,unit_price,discount_%
0,ITEM-000001,ORD-10000,PROD-04,4.0,120,15.0
1,ITEM-000002,ORD-10000,PROD-14,2.0,"99,99",20.0
2,ITEM-000003,ORD-10001,PROD-04,2.0,"119,99",20.0
3,ITEM-000004,ORD-10001,PROD-02,1.0,30,15.0
4,ITEM-000005,ORD-10001,PROD-03,1.0,34.99,0.0


In [8]:
print('=== RETURNS ===')
display(returns.head())

=== RETURNS ===


,return_id,order_id,product_id,return_date,reason,return_status,refund_amount
0,RET-00001,ord21443,PROD-04,2024/07/15,Size issue,approved,105.58
1,RET-00002,ORD-11906,PROD-05,2023-12-24,DEFECTIVE,APPROVED,119.98
2,RET-00003,13328,PROD-10,2024-11-16,Size issue,APPROVED,51.58
3,RET-00004,ORD-11511,PROD-03,2024-09-20,Damaged,Approved,29.66
4,RET-00005,ORD-13610,PROD-03,20-12-2023,NOT AS DESCRIBED,Accepted,52.48


## Step 5: Data Types & Info
`.info()` shows column names, data types, and non-null counts.
This is the most important first step — tells us what needs cleaning.

In [9]:
for name, df in tables.items():
    print(f'\n=== {name.upper()} ===')
    df.info()


=== CUSTOMERS ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4500 entries, 0 to 4499
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    4500 non-null   object
 1   customer_name  4485 non-null   object
 2   email          4500 non-null   object
 3   phone          4459 non-null   object
 4   state          4500 non-null   object
 5   signup_date    4500 non-null   object
 6   segment        4500 non-null   object
dtypes: object(7)
memory usage: 246.2+ KB

=== PRODUCTS ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    15 non-null     object 
 1   product_name  15 non-null     object 
 2   category      15 non-null     object 
 3   cost_price    14 non-null     float64
 4   retail_price  15 non-null     float64
 5   stock_qty     14 non

## Step 6: Missing Values Check
`.isnull().sum()` counts missing values per column.
Like counting NULLs in SQL — we saw these numbers in Script 01!

In [10]:
print('Missing values per table:\n')
for name, df in tables.items():
    nulls = df.isnull().sum()
    nulls = nulls[nulls > 0]  # only show columns that have nulls
    print(f'--- {name} ---')
    if len(nulls) == 0:
        print('  No missing values!')
    else:
        print(nulls.to_string())
    print()

Missing values per table:

--- customers ---
customer_name    15
phone            41

--- products ---
cost_price     1
stock_qty      1
description    4

--- orders ---
shipping_days     129
notes            5971

--- order_items ---
quantity       264
unit_price     540
discount_%    3382

--- returns ---
refund_amount    76



## Step 7: Check Dirty Values
Look at unique values in key columns to spot inconsistencies.
Like `SELECT DISTINCT` in SQL.

In [11]:
# Check segment values in customers
print('Segment values:')
print(customers['segment'].value_counts())

Segment values:
segment
OCCASIONAL    419
Occ           361
Occasional    359
occasional    356
occ           342
loyal         303
LOYAL         274
Lyl           271
Loyal         259
NEW           239
New           221
new           215
n             212
N             207
VIP           125
Vip           119
V.I.P         109
vip           109
Name: count, dtype: int64


In [12]:
# Check order_status values in orders
print('Order status values:')
print(orders['order_status'].value_counts())

Order status values:
order_status
Completed      1373
Done           1358
Complete       1336
COMPLETED      1300
completed      1280
DONE           1242
Pending         455
pending         454
In Progress     442
PENDING         436
Canceled        310
Cancelled       310
Refunded        297
cancelled       288
Refund          285
REFUNDED        284
CANCELLED       277
refunded        273
Name: count, dtype: int64


In [13]:
# Check payment_method values in orders
print('Payment method values:')
print(orders['payment_method'].value_counts())

Payment method values:
payment_method
Debit Card     644
shop pay       609
Apple Pay      603
DEBIT          600
Shop Pay       595
debit card     587
apple pay      583
Debit          576
APPLE PAY      576
ApplePay       573
SHOP PAY       571
ShopPay        561
paypal         514
pp             498
PAYPAL         490
Pay Pal        489
PayPal         460
Credit Card    430
cc             427
CreditCard     425
CC             411
CREDIT         399
credit card    379
Name: count, dtype: int64


In [14]:
# Check product names
print('Product names:')
print(products['product_name'].value_counts())

Product names:
product_name
AIR HOODIE              1
Classic T               1
Drrop Tee               1
CARGO PANTS             1
JOGGER SLIM             1
Pixel Cap Snapback      1
BUCKET HAT              1
retro runner sneaker    1
Hightop Force           1
SlidePro                1
Chain Neklace           1
CANVAS TOTE             1
Puffa Jacket            1
zip hoodie              1
graphic tee vol2        1
Name: count, dtype: int64


## Step 8: Check Numeric Issues
Look for bad quantities, impossible discounts, and price format issues.

In [15]:
# Check quantity column - convert to numeric first
order_items['quantity_check'] = pd.to_numeric(order_items['quantity'], errors='coerce')

bad_qty = order_items[order_items['quantity_check'] <= 0]
null_qty = order_items[order_items['quantity_check'].isnull()]

print(f'Bad quantities (<=0): {len(bad_qty)}')
print(f'Null quantities: {len(null_qty)}')

# Drop the temp column
order_items.drop(columns=['quantity_check'], inplace=True)

Bad quantities (<=0): 562
Null quantities: 264


In [16]:
# Check discount column for impossible values (>100)
order_items['discount_check'] = pd.to_numeric(order_items['discount_%'], errors='coerce')

bad_discount = order_items[order_items['discount_check'] > 100]
print(f'Impossible discounts (>100%): {len(bad_discount)}')
print(f'Sample bad discounts:')
print(bad_discount[['item_id', 'discount_%']].head())

order_items.drop(columns=['discount_check'], inplace=True)

Impossible discounts (>100%): 566
Sample bad discounts:
         item_id  discount_%
62   ITEM-000063       126.0
77   ITEM-000078       120.0
120  ITEM-000121       107.0
234  ITEM-000235       125.0
279  ITEM-000280       140.0


## Summary — What We Found

| Issue | Table | Count |
|-------|-------|-------|
| Missing customer names | customers | 15 |
| Missing phone numbers | customers | 41 |
| Mixed date formats | orders, customers, returns | all rows |
| Inconsistent segment values | customers | 8 variants |
| Inconsistent order status | orders | 9 variants |
| Bad quantities (<=0) | order_items | ~562 |
| Impossible discounts (>100%) | order_items | ~566 |
| Mixed price formats ($, commas) | order_items | ~540 |

**Next:** Notebook 02 will clean all of this!